packages/live-share-inking/src/core/InkingManager.ts

```diff
@@ -0,0 +1,1260 @@
+                case InkingTool.highlighter:
+                case InkingTool.laserPointer:
+                    if (this._currentStroke) {
+                        this._currentStroke.addPoints(filteredPoint);
+
+                        this.notifyAddPoints(this._currentStroke.id, filteredPoint);
+                    }
+
+                    break;
+                case InkingTool.eraser:
+                    this.erase(filteredPoint);
+
+                    this.queuePointerMovedNotification(filteredPoint);
+
+                    break;
+                case InkingTool.pointEraser:
+                    this._pendingPointErasePoints.push(filteredPoint);
+
+                    this.queuePointerMovedNotification(filteredPoint);
+
+                    this.schedulePointEraseProcessing();
+
+                    break;
+            }
+        }
+    };
+
+    private onPointerUp = (e: IPointerEvent): void => {
+        const filteredPoint = this._inputFilters.filterPoint(e);
+
+        switch (this._tool) {
+            case InkingTool.pen:
+            case InkingTool.highlighter:
+            case InkingTool.laserPointer:
+                if (this._currentStroke) {
+                    this._currentStroke.addPoints(filteredPoint);
+                    this._currentStroke.end();
+
+                    this.notifyEndStroke(this._currentStroke.id, filteredPoint);
+
+                    this._currentStroke = undefined;
+                }
+
+                break;
+        }
+
+        this.flushChangeLog();
+    };
+
+    private onPointerLeave = (e: IPointerEvent): void => {
+        if (this._tool === InkingTool.laserPointer) {
+            const filteredPoint = this._inputFilters.filterPoint(e);
+
+            this.cancelCurrentStroke(filteredPoint);
+        }
+        else {
+            this.queuePointerMovedNotification();
+        }
+    }
+
+    private internalAddStroke(stroke: IStroke, options?: IAddRemoveStrokeOptions) {
+        const effectiveOptions: Required<IAddRemoveStrokeOptions> = {
+            forceReRender: options ? (options.forceReRender ?? false) : false,
+            addToChangeLog: options ? (options.addToChangeLog ?? true) : true
+        };
+
+        if (effectiveOptions.forceReRender || this._strokes.has(stroke.id)) {
+            this._strokes.set(stroke.id, stroke);
+
+            effectiveOptions.forceReRender ? this.reRender() : this.scheduleReRender();
+        }
+        else {
+            this._strokes.set(stroke.id, stroke);
+
+            if (!this._reRenderTimeout) {
+                this._dryCanvas.renderStroke(stroke);
+            }
+        }
+
+        if (effectiveOptions.addToChangeLog) {
+            this._changeLog.addStroke(stroke);
+        }
+    }
+
+    private wetStrokeEnded(stroke: IWetStroke, isCancelled: boolean) {
+        if (!isCancelled) {
+            if (stroke.type === StrokeType.ephemeral) {
+                const effectiveClientId = stroke.clientId ?? InkingManager.localClientId;
```

Comment:
Does effectiveClientId mean 'client id of the person making the wet stroke'? I kind of like `strokeOwnerClientId`

Reply:
I use the term "effective" when I have this type of logic:
- I have a variable that may contain a value
- If it doesn't I have a fallback
- And I want a variable that holds the value I will "effectively" use in the code that follows, i.e. either the value of the variable if provided, the value of the fallback otherwise

So it doesn't represent anything beyond a clientId, it's just the one I'm effectively using in the code that follows.

Hopefully you're fine with me keeping it that way.